Sources/Support/StoreMessagesHelper.swift

```diff
@@ -0,0 +1,187 @@
+//
+//  Copyright RevenueCat Inc. All Rights Reserved.
+//
+//  Licensed under the MIT License (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      https://opensource.org/licenses/MIT
+//
+//  StoreMessagesHelper.swift
+//
+//  Created by Antonio Rico Diez on 27/9/23.
+
+import StoreKit
+
+class StoreMessagesHelper {
+
+    private let systemInfo: SystemInfo
+    private let showStoreMessagesAutomatically: Bool
+    private let storeMessagesProvider: StoreMessagesProvider
+
+    // WIP: Need to see how this handles concurrency
+    private var deferredMessages: [StoreMessage] = []
+
+    init(systemInfo: SystemInfo,
+         showStoreMessagesAutomatically: Bool) {
+        self.systemInfo = systemInfo
+        self.showStoreMessagesAutomatically = showStoreMessagesAutomatically
+        self.storeMessagesProvider = StoreMessagesProviderWrapper()
+    }
+
+    @available(iOS 16.0, *)
+    init(systemInfo: SystemInfo,
+         showStoreMessagesAutomatically: Bool,
+         storeMessagesProvider: StoreMessagesProvider = StoreMessagesProviderWrapper()) {
+        self.systemInfo = systemInfo
+        self.showStoreMessagesAutomatically = showStoreMessagesAutomatically
+        self.storeMessagesProvider = storeMessagesProvider
+    }
+
+    #if os(iOS)
+
+    @available(iOS 16.4, *)
+    @available(macOS, unavailable)
+    @available(watchOS, unavailable)
+    @available(tvOS, unavailable)
+    func deferMessagesIfNeeded() async throws {
+        guard !self.showStoreMessagesAutomatically else {
+            return
+        }
+        Task(priority: .background) {
+            for try await message in self.storeMessagesProvider.messages {
+                self.deferredMessages.append(message)
+            }
+        }
+    }
+
+    @available(iOS 16.4, *)
+    @available(macOS, unavailable)
+    @available(watchOS, unavailable)
+    @available(tvOS, unavailable)
+    @MainActor
+    func showStoreMessages(types: Set<StoreMessageType>) {
+        for message in self.deferredMessages {
+            if let messageType = message.reason.messageType {
+                if types.contains(messageType) {
+                    do {
+                        try message.display(in: self.systemInfo.currentWindowScene)
+                    } catch {
+                        Logger.error("Error displaying StoreKit message: \(error)")
+                    }
+                }
+            }
+        }
+    }
+
+    #endif
+}
+
+// @unchecked because:
+// - Class is not `final` (it's mocked). This implicitly makes subclasses `Sendable` even if they're not thread-safe.
+// - It has mutable `_deferredMessages` which is necessary due to the availability annotations.
+extension StoreMessagesHelper: @unchecked Sendable {}
+
+protocol StoreMessagesProvider {
```

Comment:
Had to move a lot of stuff behind protocols in order to be able to test it, which made the code much more complex IMO... Honestly, I kinda miss `mockk` even with its defects.

Reply:
hoping that macros and the new testing direction eventually solve this